File: leetcode-java/array/84. Largest Rectangle in Histogram.java
package array;
class Solution11 {
    public int largestRectangleArea(int[] heights) {
        int ans = Integer.MIN_VALUE;
        if(heights == null || heights.length == 0) return 0;
        int [] lessFromLeft = new int[heights.length];
        int [] lessFromRight = new int[heights.length];
        lessFromRight[heights.length - 1] = heights.length;
        lessFromLeft[0] = -1;
        for(int i = 1; i < heights.length; i ++) {
            int p = i - 1;
            while(p >= 0 && heights[i] <= heights[p]) {
                p = lessFromLeft[p];
            }
            lessFromLeft[i] = p;
        }

        for(int i = heights.length - 2; i >= 0; i --) {
            int p = i + 1;
            while(p < heights.length && heights[i] <= heights[p]) {
                p = lessFromRight[p];
            }
            lessFromRight[i] = p;
        }
        for(int i = 0; i < heights.length; i ++) {
            ans = Math.max(ans, heights[i] * (lessFromRight[i] - lessFromLeft[i] - 1));
        }
        return ans;
    }
    public static void main(String[] args) {
        Solution11 s = new Solution11();
        int [] input = {2,1,5,6,2,3};
        s.largestRectangleArea(input);
    }
}

File: leetcode-js/array/84. Largest Rectangle in Histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    heights.push(0);
    let stack = [];
    let ans = 0;
    for(let i = 0; i < heights.length; i ++) {
        while(stack.length != 0 && heights[stack[stack.length - 1]] > heights[i]) {
            let top = stack.pop();
            ans = Math.max(ans, heights[top] * (stack.length == 0 ? i : i - stack[stack.length - 1] - 1));
        }
        stack.push(i);
    }
    return ans;
};

largestRectangleArea([2,1,5,6,2,3])